iOS content for Google Authenticator (#3148)
okta/okta-developer-docs: packages/@okta/vuepress-site/docs/guides/authenticators-google-authenticator/main/ios/challengeintegrationsteps.md

### 1 - 4: Sign-in and Select Authenticator
The challenge flow follows the same first 3-4 steps as the [enrollment flow](/docs/guides/authenticators-google-authenticator/ios/main/#integrate-sdk-for-authenticator-enrollment):

* Display the initial sign-in UI
* Start the sign-on flow
* Authenticate the user credentials

> * Display a list of possible authenticator factors

Author: Good catch of clumsy wording. It actually is the list of authenticators in which the user can enroll rather than ones that have already enrolled in. How about:

> Display the list of authenticator choices

> Step 4, displaying a list of authenticators, depends on the policy settings.

Author: Good point, I was trying to take care of the general case. Line 9 deleted.

### 5: Display and capture a code entry

The `.challengeAuthenticator` remediation type used for requesting a password is also used for one-time codes. A Google Authencator code request is represented by an authenticator with a `type` of `.app` and a `key` of `"google_otp"`.

This version of `handleChallenge(remediation:)` includes Google Authenticator:

```swift
func handleChallenge(_ remediation: Remediation) {
    guard let authenticator = remediation.authenticators.first else {
        finish(with: .unexpectedAuthenticator)
        return
    }

    let type = authenticator.type
    switch type {
    case .password:
        remediation["credentials.passcode"]?.value = password
        remediation.proceed()

    case .app:
        if authenticator.key == "google_otp" {
            // Tell the UI to present a UI for requesting a one-time key.
        }

    default:
        finish(with: .unexpectedAuthenticator)
        return
    }
}
```

> The same utlity functions work for both a challenge and an enrollment. The SwiftUI form also works because the QR Code and shared secret display only if there's data. The next steps are also the same, either a successful login, a message, or an error.

Author: Good catch on unclear wording. I'm talking about how the app handles the server response to the current form. It could be a successful login (end of the flow), a message (e.g. incorrect code), or some error that terminates the process (e.g. unable to reach the server). How about:

> The next response from the SDK are also handled the same way: a successful login ends the sign-on flow, a message is displayed (e.g. an incorrect code), and an error terminates the sign-on flow.

> The same utlity functions work for both a challenge and an enrollment. The SwiftUI form also works because the QR Code and shared secret display only if there's data. The next steps are also the same, either a successful login, a message, or an error.

Reviewer: That makes sense.  Not sure if it should be "response is or "responses are" in first phrase....

> The same utlity functions work for both a challenge and an enrollment. The SwiftUI form also works because the QR Code and shared secret display only if there's data. The next steps are also the same, either a successful login, a message, or an error.

Reviewer: nice

* [Authenticate the user credentials](#_3-authenticate-the-user-credentials).
* [Display a list of authenticators available for enrollment](#_4-display-a-list-of-authenticators-available-for-enrollment).

### 5: Display an OTP entry field

The `.challengeAuthenticator` remediation type used for requesting a password is also used for one-time codes. A Google Authencator code request is represented by an authenticator with a `type` of `.app` and a `key` of `"google_otp"`.

This version of `handleChallenge(remediation:)` includes Google Authenticator:

```swift
func handleChallenge(_ remediation: Remediation) {
    guard let authenticator = remediation.authenticators.first else {
        finish(with: .unexpectedAuthenticator)
        return
    }

    let type = authenticator.type
    switch type {
    case .password:
        remediation["credentials.passcode"]?.value = password
        remediation.proceed()

    case .app:
        if authenticator.key == "google_otp" {
            // Tell the UI to present a UI for requesting a one-time key.
        }

    default:
        finish(with: .unexpectedAuthenticator)
        return
    }
}
```

> The `EnrollAuthenticatorView` from [the enrollment flow](#_5-display-the-shared-secret-qr-code-and-request-the-code) also works for the challenge case as the shared secret section is shown only if the data is in the remediation. The following screenshots show that view with a field for entering the OTP, and the same view display a message that's part of the remediation.

Reviewer 2: ```suggestion
The `EnrollAuthenticatorView` from [the enrollment flow](#_5-display-the-shared-secret-qr-code-and-request-the-code) also works for the challenge case as the shared secret section is shown only if the data is in the remediation. The following screenshots show that view with a field for entering the OTP, and the same view displays a message that's part of the remediation.
```

Author: First, I'm not sure how to comment on the flow diagram comments. I'll update the wording as appropriate. The diagrams themselves are draft version form which Alex will create final ones. I think all the text is copy/pasted from the .net flowcharts and then modified with the iOS specific info.

I'll do a revision for the next commit (likely late Monday).

For this item, thanks for catching the typo, the fix will be in next commit.

<div class="common-image-format">

![Display the OTP entry field and any warning message.](/img/authenticators/ios-authenticators-google-code-screen.png "Two screenshots. The first shows a form containing an OTP code entry filed and continue and cancel buttons. The second shows the same form with a warning message.")

</div>

> The next response from the SDK is also handled the same way, and is either: a successful login that ends the sign-on flow, a message to display (e.g. an incorrect code), or an error that terminates the sign-on flow.

Reviewer 2: style is the use of sign in, sign-in and so on, not sign-on or sign on - unless it's used in a UI somewhere....
Also avoid the use of sign in as a noun

```suggestion
The next response from the SDK is also handled the same way, and is either: a user who successfully signs in and ends the sign-in flow, a message to display (for example, an incorrect code), or an error that terminates the sign-in flow.
```

Author: Updates will be in the next commit.
